Upgrading to python-wink 0.7.4 and improving RGB color support in HA

bradsk88 opened the pull request:
**Description:**

Upgrades wink to 0.7.4
Makes RBG color and color temperature work better.  Prior to this commit, only bulbs which supported XY color would work.  Now color temperature, RGB, and HSB work as they should.

**Checklist:**

If code communicates with devices:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [ ] ~~New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51))~~.
- [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [ ] ~~New files were added to `.coveragerc`.~~

jaharkes commented on homeassistant/components/light/wink.py:
```diff
@@ -63,20 +71,48 @@ def available(self):
+            state_kwargs['color_xy'] = color_util.color_RGB_to_xy(*rgb_color)
+
+        if color_temp_mired:
+            # HA returns mireds in reverse.  Flip it around.
```
Odd, the temperature value in mireds seemed just fine when I added color support for WeMo lights.

bradsk88 replied:
The max value of the color temp slider is 500 and the minimum is ~150

kelvin = 1 million / mired
1M / 500 = 2000K (orange-ish)
1M / 150 = 6666K (blue-ish)

jaharkes replied:
But that is how it is for all other lights that support temperature in HA, so as far as home-assistant is concerned it is the conventional style. Also when scripting, people will expect that giving a temperature value 400 mireds (or 2500K) results in a warm white.

If you don't like the direction of the slider, this is not the place to fix it.

bradsk88 replied:
I'm not "fixing" it.  I'm doing what every other person who has used the values has done.  I'm bending them to fit my application.

balloob replied:
I agree with @jaharkes. If our UI is the opposite of the conventional style and thus sends out values that have to be reversed, it should be fixed in the main UI. So I also would like to see line 104 is removed and the more info light dialog is fixed.

bradsk88 replied:
I can certainly do that, but all other light bulb implementations will be flip-flopped.
Fair warning.

bradsk88 replied:
@balloob On that note.  Do you know if there was any reason the minimum of 154 and maximum of 500 were chosen?  

I feel like that range could, practically, be expanded to 800 and 100 because many color-flexible bulbs will operate in those ranges.

balloob commented on homeassistant/components/light/__init__.py:
```diff
@@ -288,10 +288,20 @@ def state_attributes(self):
-                data[ATTR_RGB_COLOR] = color_util.color_xy_brightness_to_RGB(
-                    data[ATTR_XY_COLOR][0], data[ATTR_XY_COLOR][1],
-                    data[ATTR_BRIGHTNESS])
+            data[ATTR_RGB_COLOR] = self._get_light_rgb_color(data)
```
Why did you extract this into a method? It does not seem to be used by anything else. Please revert this.

bradsk88 replied:
Made it easier to fit in the 79 character limit.
I'll find another way.

balloob commented on homeassistant/util/color.py:
```diff
@@ -90,3 +91,92 @@ def rgb_hex_to_rgb_list(hex_string):
+    return blue
+
+
+HASS_COLOR_MAX = 500  # mireds (inverted)
```
Please keep constants at the top of the file.

balloob commented on homeassistant/util/color.py:
```diff
@@ -90,3 +91,92 @@ def rgb_hex_to_rgb_list(hex_string):
+    if temperature <= 66:
+        return 255
+    tmp_red = 329.698727446 * math.pow(temperature - 60, -0.1332047592)
+    if tmp_red < 0:
```
```python
return max(0, min(255, red))
```

balloob commented:
This will require some tests for the newly added color methods so our test coverage won't decrease: https://coveralls.io/builds/5819580/source?filename=homeassistant%2Futil%2Fcolor.py

bradsk88 commented:
@balloob I think I've addressed all of your comments.  The TravisCI failure appears to be unrelated to my changes.

I'll look into inverting the color temperature slider in another PR

balloob commented:
The Travis CI failures are related to your changes. They are linting violations.

balloob commented:
Not sure why Farcy didn't run but I just triggered a build.

balloob commented on homeassistant/util/color.py:
```diff
@@ -90,3 +94,73 @@ def rgb_hex_to_rgb_list(hex_string):
+    return (red, green, blue)
+
+
+def _bound(color_component, minimum=0, maximum=255):
```
I'm not sure what the usefulness is of making this an extra method. Why not just inline the logic

bradsk88 replied:
Just avoids duplicating 2 lines 3 times.  It's a minor implementation detail.

jaharkes commented:
Should farcy complain about missing docstrings on test functions and classes? 

On April 17, 2016 6:04:21 PM EDT, Paulus Schoutsen <notifications@github.com> wrote:
>Not sure why Farcy didn't run but I just triggered a build.
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/home-assistant/home-assistant/pull/1832#issuecomment-211122347

balloob commented:
@jaharkes Farcy only validates new code. As we eventually want to get full pep257 compliant test code, making sure new code fits the bill is a good guard.

bradsk88 commented:
Given that my test names are already fully descriptive, I have no interest in writing docstrings for every test method.

Is that *seriously* a requirement of PEP257?

bradsk88 commented:
Follow-up:

Looking into it, I see that PEP257 is just a set of recommendations for docstrings in general.  I find it questionable that you're choosing to applying that to tests - given how much of a chore test writing can be in the first place.

balloob commented:
It might seem like a chore to you but its one of the few measurements that
I have put in place to keep this project manageable. There is a lot of code
in home Assistant and these tests ensure that we can go around and improve
things without having to be afraid that something gets left in a broken
state.

Code will be always read more often than written. The names might make
sense for you now but they might not make sense to the next person who only
sees the test fail related to another change.

On Sun, Apr 17, 2016, 15:21 Brad Johnson <notifications@github.com> wrote:

> Follow-up:
>
> Looking into it, I see that PEP257 is just a set of recommendations for
> docstrings in general. I find it questionable that you're choosing to
> applying that to tests - given how much of a chore test writing can be in
> the first place.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/1832#issuecomment-211123956>
>

bradsk88 commented:
And who is going to be in charge of ensuring that the test docstrings don't rot?

bradsk88 commented on tests/util/test_color.py:
```diff
@@ -57,3 +57,54 @@ def test_rgb_hex_to_rgb_list(self):
+        self.assertEqual(200, mired)
+
+
+class ColorTemperatureToRGB(unittest.TestCase):
```
@balloob Farcy should comment one line lower so that, when I actually add the docstring, this comment can be auto-hidden.

bradsk88 commented:
Docstrings added

balloob commented on homeassistant/components/frontend/www_static/home-assistant-polymer:
```diff
@@ -1 +1 @@
-Subproject commit 1ad673ff6f946de2a6358efcdf59667fbc73a05a
+Subproject commit a21547abfb8d19605850ec4bee48f837d607f736
```
Can you remove this from your commit. This has to go together with a new build of the frontend.

bradsk88 replied:
Can this be added to .gitignore?

balloob replied:
No, because it is needed for when we add a new compiled version of the frontend to the main repo.

balloob commented:
Thanks for addressing the issues Brad. I really appreciate your fixes even though you might not agree with them.

Commit:
Upgrading to python-wink 0.7.4 and improving RGB color support in HA

bradsk88 commented:
Maintainer is always right :grinning:
